client/my-sites/checkout/src/components/wp-order-review-line-items.tsx

```diff
@@ -127,6 +139,22 @@ export function WPOrderReviewLineItems( {
 		);
 	}, [ responseCart.products ] );
 
+	useEffect( () => {
+		const productOrderNeedsInitialization = ! cartProductsOrder.length;
+
+		// An update can be either a new product added to the cart or an existing product
+		// that has been updated (e.g. changing the duration of a plan).
+		// Removing a product from the cart will not trigger this effect, since the
+		// removed product will then be shown in the restorable products list.
+		const cartHasBeenUpdated = cartProductsOrder.length <= responseCart.products.length;
+
+		if ( productOrderNeedsInitialization || cartHasBeenUpdated ) {
+			setCartProductsOrder( responseCart.products.map( ( { product_id } ) => product_id ) );
+		}
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps -- Having cartProductsOrder as a dependency would trigger an infinite loop.
```

Comment:
You can use a callback to `setCartProductsOrder()` instead of using `cartProductsOrder` directly. That will be more reliable and not require this hack. You can put the `productOrderNeedsInitialization || cartHasBeenUpdated` conditions inside that callback.